### PR TITLE
Permit BC apps to choose their cluster

### PR DESCRIPTION
**Pull request by MorganRodgers**
Resurfacing @ericfranz's work.

For history/notes see: https://github.com/OSC/ood-dashboard/pull/424

### CODE REVIEW

File: apps/dashboard/app/controllers/batch_connect/session_contexts_controller.rb [hunk 1]
```diff
@@ -23,19 +23,21 @@ def new
...
 
     @session = BatchConnect::Session.new
     respond_to do |format|
-      if @session.save(app: @app, context: @session_context, format: @render_format)
+      if @cluster && @session.save(app: @app, context: @session_context, format: @cluster.job_config[:adapter], cluster_id: @cluster.id)
         cache_file.write(@session_context.to_json)  # save context to cache file
         format.html { redirect_to batch_connect_sessions_url, notice: t('dashboard.batch_connect_sessions_status_blurb_create_success') }
         format.json { head :no_content }
       else
+        @session.errors.add(:base, :cluster_not_specified, "A valid cluster to submit to has not been specified.") unless @cluster
```
Comment (MorganRodgers) [inline apps/dashboard/app/controllers/batch_connect/session_contexts_controller.rb]: Crashes with a NoMethodError when the requested cluster is not available:

```
undefined method `except' for "A valid cluster to submit to has not been specified.":String
```

Comment (MorganRodgers) [reply apps/dashboard/app/controllers/batch_connect/session_contexts_controller.rb]: Fixed in 41c64df.

**Comment by MorganRodgers** [comment]
When using this branch I can successfully submit a demo app to either Pitzer or Owens using the following `form.yml`:

```yaml
---
attributes:
    bc_account:
        help: You can leave this blank if **not** in multiple projects.
    bc_queue: null
    bc_vnc_idle: 0
    bc_vnc_resolution:
        required: true
    desktop:
        label: Ansol Desktop environment
        options:
        - [ "Xfce", "xfce" ]
        - [ "Mate", "mate" ]
        widget: select
    cluster:
        label: Cluster
        options:
        - [ "Pitzer, but faster", "quick_pitzer" ]
        - [ "Owens", "owens" ]
        widget: select
cluster:
  - "quick_pitzer"
  - "owens"
form:
- cluster
- bc_vnc_idle
- desktop
- bc_account
- bc_num_hours
- bc_queue
- bc_vnc_resolution
- bc_email_on_started
title: Pitzer Desktop Test
```

**Comment by MorganRodgers** [comment]
I have confirmed that this works with heterogeneous schedulers. There are still concerns about how it behaves when one or more clusters is either hidden or access controlled (using file permissions on the cluster config or the ACL).

**Comment by johrstrom** [comment]
I'd vote for this being a part of 1.7 for sure (someone just asked about it on discourse). 

I can replicate the behaviour Morgan describes, but if I list a cluster that I can't use (a file with 200 permissions that I can't read) I don't even see the app in the dashboard list, so there is something broken there.

**Comment by baverhey** [comment]
@johrstrom: Can we tackle this?
I tried to apply this to 1.7.10
#<ActionView::Template::Error: undefined method `staged_root' for nil:NilClass>

**Comment by johrstrom** [comment]
@baverhey I'm unaware of that issue, I guess I'd have to see the stack trace.

But as to priority, yes I've put the original ticket #227 in the 1.8 milestone so we can keep track of it's progress there.

**Comment by ericfranz** [comment]
Duplicate PR of https://github.com/OSC/ondemand/pull/524

**Pull request closed without merging**